source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,702 @@
+  }
+
+  // Construct a new composite.
+  uint32_t compositeId = TakeNextId();
+  where = load;
+  std::unique_ptr<ir::Instruction> compositeConstruct(new ir::Instruction(
+      context(), SpvOpCompositeConstruct, load->type_id(), compositeId, {}));
+  for (auto l : loads) {
+    ir::Operand op(SPV_OPERAND_TYPE_ID,
+                   std::initializer_list<uint32_t>{l->result_id()});
+    compositeConstruct->AddOperand(std::move(op));
+  }
+  where = where.InsertBefore(std::move(compositeConstruct));
+  get_def_use_mgr()->AnalyzeInstDefUse(&*where);
+  context()->set_instr_block(&*where, block);
+  context()->ReplaceAllUsesWith(load->result_id(), compositeId);
+}
+
+void ScalarReplacementPass::ReplaceWholeStore(
+    ir::Instruction* store, const std::vector<ir::Instruction*>& replacements) {
+  // Replaces a store to the whole composite with a series of extract and stores
+  // to each element.
+  uint32_t storeInput = store->GetSingleWordInOperand(1u);
+  ir::BasicBlock* block = context()->get_instr_block(store);
+  ir::BasicBlock::iterator where(store);
+  uint32_t elementIndex = 0;
+  for (auto var : replacements) {
+    // Create the extract.
+    ir::Instruction* type = GetStorageType(var);
+    uint32_t extractId = TakeNextId();
+    std::unique_ptr<ir::Instruction> extract(new ir::Instruction(
+        context(), SpvOpCompositeExtract, type->result_id(), extractId,
+        std::initializer_list<ir::Operand>{
+            {SPV_OPERAND_TYPE_ID, {storeInput}},
+            {SPV_OPERAND_TYPE_LITERAL_INTEGER, {elementIndex++}}}));
+    auto iter = where.InsertBefore(std::move(extract));
+    get_def_use_mgr()->AnalyzeInstDefUse(&*iter);
+    context()->set_instr_block(&*iter, block);
+
+    // Create the store.
+    std::unique_ptr<ir::Instruction> newStore(
+        new ir::Instruction(context(), SpvOpStore, 0, 0,
+                            std::initializer_list<ir::Operand>{
+                                {SPV_OPERAND_TYPE_ID, {var->result_id()}},
+                                {SPV_OPERAND_TYPE_ID, {extractId}}}));
+    // Copy memory access attributes.
+    for (uint32_t i = 2; i < store->NumInOperands(); ++i) {
+      ir::Operand copy(store->GetInOperand(i));
+      newStore->AddOperand(std::move(copy));
+    }
+    iter = where.InsertBefore(std::move(newStore));
+    get_def_use_mgr()->AnalyzeInstDefUse(&*iter);
+    context()->set_instr_block(&*iter, block);
+  }
+}
+
+bool ScalarReplacementPass::ReplaceAccessChain(
+    ir::Instruction* chain, const std::vector<ir::Instruction*>& replacements) {
+  // Replaces the access chain with either another access chain (with one fewer
+  // indexes) or a direct use of the replacement variable.
+  uint32_t indexId = chain->GetSingleWordInOperand(1u);
+  const ir::Instruction* index = get_def_use_mgr()->GetDef(indexId);
+  size_t indexValue = GetConstantInteger(index);
+  if (indexValue > replacements.size()) {
+    // Out of bounds access, this is illegal IR.
+    return false;
```

Comment:
Easier to just assert?  Nothing good will happen from this point on.

Reply:
This causes the pass to return Failure, which I think is better than only catching it in debug.